pkg/env/certificateRefresher_test.go

```diff
@@ -0,0 +1,293 @@
+MRIwEAYDVQQDEwlwZXRyLnRlc3QwHhcNMjEwNTI2MDkzNDQzWhcNMjEwNjI2MDk0
+NDQzWjAUMRIwEAYDVQQDEwlwZXRyLnRlc3QwggEiMA0GCSqGSIb3DQEBAQUAA4IB
+DwAwggEKAoIBAQDGwJS2bDR2ugfsODr1/VkhX9A2b6/w66r4habosbIeX9meMsKZ
+btsMOrWH2uten9H7L0o2rYbU5c3LAtEaeEplGG4AamtF4s35axIF8/jgMvkswHjo
+db2f79iImOx96tE/Odhu3T/uoR1V543Q1f80/HZzdpqnIpNZd/NrDMv2Hjm1LcLF
+MpcZFBpdF4CmAm6J59LjfZWm5qq5aJf3MFEdxsX0VwEURmP0lrsvZTJhsu+httv+
+soFareiOk2W5nU0UhFSpKfOrFfldcHxNPX6T2wr7JTYLDfgeRG6VoboPwcdVCCll
+rKWUlP9zdq+A6zbByHqC+QG6IP2czzXUAMR9AgMBAAGjfDB6MA4GA1UdDwEB/wQE
+AwIFoDAJBgNVHRMEAjAAMB0GA1UdJQQWMBQGCCsGAQUFBwMBBggrBgEFBQcDAjAf
+BgNVHSMEGDAWgBTVoGOzKcn+Y+vypnh2e5BzIcQ6NzAdBgNVHQ4EFgQU1aBjsynJ
+/mPr8qZ4dnuQcyHEOjcwDQYJKoZIhvcNAQELBQADggEBAGREAWRpsAUlhLtnqgyh
+qhRDd0xuzlBte5uMWYJxojwK7jmJlb4YbIgm6OfSMW3cd1xN497egji5AiiW37wN
+UDONqYUL4qF3nnPJl9yjvptJCdpZQC1Ma5b2BLF8X8kurh8rYtyx2NNncj6/p/IA
+D/R4qcuA0UaXLdARsppjLTCzk6T/Z8v4qaaXQJBfdHLY2fZo4vSI1yLyTcGia703
+Zn30eMz0PLFCkMEYhZADUia0peReH08WO7RTtE5Nf8r5ZiFSfuJJlBmURMU3MAxJ
+PeihXgcD511I3eMq7A8r7+xlNFEFYaleYwmoYSOt0cBGtoZttXDYacR/xvEkVaxO
+1LQ=
+-----END CERTIFICATE-----`
+
+func equalPrivKey(one, other *rsa.PrivateKey) bool {
+	if one == nil && other == nil {
+		return true
+	}
+
+	pubEqual := one.PublicKey.N.Cmp(other.PublicKey.N) == 0 && one.PublicKey.E == other.PublicKey.E
+	expEqual := one.D.Cmp(other.D) == 0
+	if !pubEqual || !expEqual {
+		return false
+	}
+
+	primesEqual := len(one.Primes) == len(other.Primes)
+	if !primesEqual {
+		return false
+	}
+
+	for i := range one.Primes {
+		if one.Primes[i].Cmp(other.Primes[i]) != 0 {
+			return false
+		}
+	}
+
+	return true
+}
+
+func TestRefreshingCertificate(t *testing.T) {
+
+	key1, certs1, err := utilpem.Parse([]byte(testCertBundle1))
+	if err != nil {
+		return
+	}
+
+	key2, certs2, err := utilpem.Parse([]byte(testCertBundle2))
+	if err != nil {
+		return
+	}
+
+	testCertName := "rotation-test-0"
+
+	type certTuple struct {
+		privateKey *rsa.PrivateKey
+		certs      []*x509.Certificate
+		err        error
+	}
+
+	cannotPull := errors.New("Cannot pull certificates from keyvault")
+
+	getterFactory := func(ct []certTuple) CertGetter {
+		// mocks the KeyvaultManager.GetCertificateSecret
+		i := 0
+		return func(c context.Context, s string) (*rsa.PrivateKey, []*x509.Certificate, error) {
+			defer func() {
+				if i < len(ct)-1 {
+					i++
+				}
+			}()
+			return ct[i].privateKey, ct[i].certs, ct[i].err
+		}
+	}
+
+	tt := []struct {
+		name       string
+		interval   time.Duration
+		sleep      time.Duration
+		getter     CertGetter
+		wantKey    *rsa.PrivateKey
+		wantCert   *x509.Certificate
+		wantErr    error
+		wantUpdate bool
+	}{
+		{
+			name:     "test initial certificate, timer set to long time",
+			interval: 1 * time.Minute,
+			sleep:    0 * time.Millisecond,
+			getter: getterFactory([]certTuple{
+				{
+					privateKey: key1,
+					certs:      certs1,
+					err:        nil,
+				},
+			}),
+			wantKey:  key1,
+			wantCert: certs1[0],
+			wantErr:  nil,
```

Comment:
Zero value of `error` is `nil`. You can remove this line.

Reply:
Yes, I know the zero value is nil.  I prefer explicit over implicit, as this is easier for everyone after me to at first glance to see my intent.